### PR TITLE
Add WebNN-native binding for Node.js to expose WebNN JavaScript API

### PR DESCRIPTION
**Pull request by fujunwei**
[webnn-polyfill](https://webmachinelearning.github.io/webnn-polyfill/test/) has 600 test cases to test APIs, ops and models. To avoid porting of these tests to C++, it would be useful to run these JS tests with node.js. The WebNN-native binding is to bind WebNN-native C++ API with addon.

This PR depends on https://github.com/webmachinelearning/webnn-native/pull/10.

### CODE REVIEW

**Comment by BruceDai** [comment]
Thanks much @fujunwei!

Current [webnn-polyfill@5c5e9f6](https://github.com/webmachinelearning/webnn-polyfill) has 615 test cases (614 passing / 1 pending by WebNN Polyfill API):
- **API tests**  total: 84, passing 84 
- **Ops tests**  total: 266, passing 266
- **Models tests** total: 15, passing 14, 1 pending
- **CTS tests** total: 250, passing 250

Following [node/README.md](https://github.com/webmachinelearning/webnn-native/blob/dad1dd6b3d2135ac0ec0e0167c11a106a53ae286/node/README.md), I verified this PR:

1. built webnn-native by DML backend, when executing **npm test** command to run all tests of webnn-polyfill, 
got **386 passing / 1 pending / 228 failing**
2. built webnn-native by OpenVINO backend, when executing **npm test** command to run all tests of webnn-polyfill, 
got **385 passing / 1 pending / 229 failing**

Let's look at pass status of running Ops tests and Model tests:
|Backend|Ops tests (npm run test-ops)|Model tests (npm run test-models)|
|----|----|----|
|DML|128 passing 138* failing|14 passing 1pending|
|OpenVINO|126 passing 140* failing|14 passing 1 pending|

Above '14 passing 1 pending' of running Model tests are same as test result by WebNN polyfill API.
And 138 failing ops tests by DML backend and 140 failing ops tests by OpenVINO backend are most with unimplemented op features likes conv2d support transpose option, and those unimplemented ops (div/exp/gru/instance_norm/max/min/pad/pow/reduce/resample/sigmoid/slice/split/squeeze/sub/tanh/transpose) by WebNN native.

WebNN polyfill Supported ops| WebNN native Supported status| 
|----|----|
add|YES|
batch_norm|YES|
clamp|YES|
concat|YES|
conv2d|YES|
div|NO|
exp|NO|
gemm|YES|
gru|NO|
instance_norm|NO|
leaky_relu|YES|
matmul|YES|
max|NO|
min|NO|
mul|YES|
pad|NO|
pool2d|YES|
pow|NO|
reduce|NO|
relu|YES|
resample|NO|
reshape|YES|
sigmoid|NO|
slice|NO|
softmax|YES|
split|NO|
squeeze|NO|
sub|NO|
tanh|NO|
transpose|NO|


Test logs:
1. [test_log_by_webnn_nodejs_binding_with_dml.txt](https://github.com/webmachinelearning/webnn-native/files/6630618/test_log_by_webnn_nodejs_binding_with_dml.txt)
2. [test_log_by_webnn_nodejs_binding_with_openvino.txt](https://github.com/webmachinelearning/webnn-native/files/6630620/test_log_by_webnn_nodejs_binding_with_openvino.txt)

File: .gitignore [hunk 1]
```diff
@@ -96,3 +96,11 @@ node/generator/specifications
 node/PATH_TO_DAWN
 node/third_party/webnn-polyfill
 node/node_modules
+node/third_party
+node/package-lock.json
```
Comment (huningxin) [inline .gitignore:100]: `package-lock.json` is intended to be checked into source control for good reason, see https://stackoverflow.com/questions/44206782/do-i-commit-the-package-lock-json-file-created-by-npm-5, perhaps we could do that too.

File: node/README.md [hunk 2]
```diff
@@ -0,0 +1,52 @@
+# WebNN-native Binding for Node.js
+
+*WebNN-native Binding for Node.js is a Node.js [C++ addon](https://nodejs.org/api/addons.html). The implementation is based on WebNN-native C API and exposes [WebNN](https://webmachinelearning.github.io/webnn/) JavaScript API.
```
Comment (huningxin) [inline node/README.md:3]: I suppose it is based on the WebNN-native C++ API now. Please correct me if I am wrong.

File: node/README.md [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+## Test
+
+```shell script
+> npm test
+```
+
+## Example
+
+ * [LeNet in Electron.js](examples/lenet_electron/README.md)
+ * [LeNet in Node.js](examples/lenet/README.md)
```
Comment (huningxin) [inline node/README.md:52]: Should the link to image classification example be added?

File: node/binding.gyp [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+              "-Wl,-rpath,'$$ORIGIN'/..",
+              '-lwebnn_native',
+              '-lwebnn_proc'
+            ],
+            'copies': [ {
+              'destination': '<(module_root_dir)/build/$(Configuration)/',
+              'files': [
+                  # webNN native.
+                  # Component build has the libc++ binary.
+                  #'<(WEBNN_NATIVE_LIB_PATH)/libc++.so',
```
Comment (huningxin) [inline node/binding.gyp:78]: Please remove the commented out code

File: node/binding.gyp [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+          'OS=="linux"', {
+            'libraries' : [
+              "-Wl,-rpath,'$$ORIGIN'/..",
+              '-lwebnn_native',
+              '-lwebnn_proc'
+            ],
+            'copies': [ {
+              'destination': '<(module_root_dir)/build/$(Configuration)/',
+              'files': [
+                  # webNN native.
```
Comment (huningxin) [inline node/binding.gyp:76]: ditto

File: node/binding.gyp [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+      'conditions': [
+        [ 'OS=="win"', {
+            'libraries' : [
+              '-lwebnn_native.dll.lib',
+              '-lwebnn_proc.dll.lib'
+            ],
+            'copies': [ {
+              'destination': '<(module_root_dir)/build/$(Configuration)/',
+              'files': [
+                  # webNN native.
```
Comment (huningxin) [inline node/binding.gyp:56]: webNN -> WebNN

File: node/binding.gyp [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+                  '<(WEBNN_NATIVE_LIB_PATH)/libc++.dll',
+                  '<(WEBNN_NATIVE_LIB_PATH)/webnn_native.dll',
+                  '<(WEBNN_NATIVE_LIB_PATH)/webnn_proc.dll',
+                  # DirectML.
+                  '<(WEBNN_NATIVE_LIB_PATH)/DirectML.dll',
+                  # OpenVINO
+                  '<(WEBNN_NATIVE_LIB_PATH)/ie_nn_c_api.dll',
```
Comment (huningxin) [inline node/binding.gyp:63]: What would happen if the webnn-native is only built for DirectML or vice versa?

File: node/examples/console/lenet/main.mjs [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
...0,0,0,0,0,0,0,0,0,0,0,5,35,207,253,254,252,164,4,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,7,92,139,249,254,254,250,112,3,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,43,206,245,251,254,254,254,248,43,1,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,2,80,239,253,254,254,254,254,251,143,3,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,5,111,239,250,250,250,253,252,168,4,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,4,74,90,91,98,234,251,170,5,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,2,2,2,7,134,250,214,34,1,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,16,245,253,208,31,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,4,2,0,0,0,0,0,4,137,251,250,117,3,0,0,0,0,0,0,0,0,0,0,0,0,0,0,25,168,91,4,1,0,1,3,34,233,253,245,40,1,0,0,0,0,0,0,0,0,0,0,0,0,0,4,156,247,210,69,39,7,54,93,201,252,250,138,11,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,60,195,248,248,218,180,235,249,253,251,205,19,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,5,133,111,247,249,250,253,254,253,226,11,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,3,12,115,118,164,245,247,229,57,1,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,3,3,4,6,6,6,1,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0]);
```
Comment (huningxin) [inline node/examples/console/lenet/main.mjs:36]: I would suggest that the lenet JS example should align to the C++ example. The [LeNet C++ example](https://github.com/webmachinelearning/webnn-native/tree/main/examples/LeNet) supports loading the .idx images.

File: node/examples/console/lenet/package.json [hunk 9]
```diff
@@ -0,0 +1,12 @@
+{
+  "name": "webnn_node_examples",
```
Comment (huningxin) [inline node/examples/console/lenet/package.json:2]: the name should be more specific, such as "webnn_node_lenet_example".

File: node/examples/electron/image_classification/main.js [hunk 10]
```diff
@@ -0,0 +1,55 @@
+// Modules to control application life and create native browser window
```
Comment (huningxin) [inline node/examples/electron/image_classification/main.js:1]: We should have a License header of this file.

File: node/examples/electron/image_classification/package.json [hunk 11]
```diff
@@ -0,0 +1,14 @@
+{
+  "name": "webnn_electron",
```
Comment (huningxin) [inline node/examples/electron/image_classification/package.json:2]: ditto, please have a more specific name

File: node/examples/electron/lenet/main.js [hunk 12]
```diff
@@ -0,0 +1,55 @@
+// Modules to control application life and create native browser window
```
Comment (huningxin) [inline node/examples/electron/lenet/main.js:1]: ditto, license header.

File: node/examples/electron/lenet/package.json [hunk 13]
```diff
@@ -0,0 +1,14 @@
+{
+  "name": "webnn_electron",
```
Comment (huningxin) [inline node/examples/electron/lenet/package.json:2]: ditto, more specific name.

File: node/examples/utils.js [hunk 14]
```diff
@@ -0,0 +1,50 @@
...
+function checkShape(shape, expected) {
+  assert.isTrue(shape.length === expected.length);
+  for (let i = 0; i < shape.length; ++i) {
+    assert.isTrue(shape[i] === expected[i]);
+  }
+}
+
+module.exports = {
+  checkValue
+}
```
Comment (huningxin) [inline node/examples/utils.js:50]: I suppose this file comes from webnn-polyfill. If it is, please use that of third_party.

File: node/examples/electron/setup.js [hunk 15]
```diff
@@ -0,0 +1,3 @@
+const webnn = require('../../lib/webnn');
+global.navigator.ml = webnn.ml;
+global.MLGraphBuilder = webnn.MLGraphBuilder;
```
Comment (huningxin) [inline node/examples/electron/setup.js:3]: should we align this one with `node_setup.js`?

File: node/src/Graph.cpp [hunk 16]
```diff
@@ -0,0 +1,282 @@
...
+                    jsTypedArray.ByteOffset());
+                resource.size = jsTypedArray.ByteLength();
+                jsElementLength = jsTypedArray.ElementSize();
+            }
+            if (HasOptionMember(jsResource, "dimensions")) {
+                if (!GetInt32Array(jsResource.Get("dimensions"), resource.dimensions)) {
+                    return false;
+                }
+
+                /***Dimensions Check ***/
```
Comment (huningxin) [inline node/src/Graph.cpp:204]: Please use the same comment style as others.

File: src/webnn_native/openvino/ienn/src/transpose_sinking.cpp [hunk 17]
```diff
@@ -0,0 +1,447 @@
+/*****************************************************************************
+ * Copyright (C) 2021 Intel Corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
```
Comment (huningxin) [inline src/webnn_native/openvino/ienn/src/transpose_sinking.cpp:4]: As we add this file, please add OpenVINO license into https://github.com/webmachinelearning/webnn-native/blob/main/NOTICE

**Review by huningxin — COMMENTED** [review]
Great work, thanks @fujunwei . I left some comments, please take a look.

**Comment by fujunwei** [comment]
@huningxin Done, thanks.

**Review by huningxin — APPROVED** [review]
Thanks for updating the PR. It looks good to me.

**Comment by huningxin** [comment]
Based on the feedback from [WebML WG Teleconference – 24 June 2021](https://www.w3.org/2021/06/24-webmachinelearning-minutes.html#t06), both the webnn-polyfill tests and webnn-native implementation on native ML APIs are useful to inform the spec development. This PR specifically enables running the webnn-polyfill tests on native ML API implementation. With that, I am going to merge this PR and unlock the development of other features, like new API change and ops. 

Thanks much @fujunwei and @BruceDai !
